x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForRolloverReadyStep.java

```diff
@@ -216,25 +224,39 @@ public void evaluateCondition(Metadata metadata, Index index, Listener listener,
 
     /**
      * Builds a RolloverRequest that captures the various max_* and min_* conditions of this {@link WaitForRolloverReadyStep}.
-     *
+     * <p>
      * To prevent empty indices from rolling over, a `min_docs: 1` condition will be injected if `rolloverOnlyIfHasDocuments` is true
      * and the request doesn't already have an associated min_docs or min_primary_shard_docs condition.
      *
-     * @param rolloverTarget the index to rollover
-     * @param masterTimeout the master timeout to use with the request
+     * @param rolloverTarget             the index to rollover
+     * @param masterTimeout              the master timeout to use with the request
      * @param rolloverOnlyIfHasDocuments whether to inject a min_docs 1 condition if there is not already a min_docs
      *                                   (or min_primary_shard_docs) condition
+     * @param targetIsTsdb               Whether the target is a tsdb data stream
      * @return A RolloverRequest suitable for passing to {@code rolloverIndex(...) }.
      */
     // visible for testing
-    RolloverRequest createRolloverRequest(String rolloverTarget, TimeValue masterTimeout, boolean rolloverOnlyIfHasDocuments) {
+    RolloverRequest createRolloverRequest(
+        String rolloverTarget,
+        TimeValue masterTimeout,
+        boolean rolloverOnlyIfHasDocuments,
+        boolean targetIsTsdb
+    ) {
         RolloverRequest rolloverRequest = new RolloverRequest(rolloverTarget, null).masterNodeTimeout(masterTimeout);
         rolloverRequest.dryRun(true);
         if (rolloverOnlyIfHasDocuments && (conditions.getMinDocs() == null && conditions.getMinPrimaryShardDocs() == null)) {
             rolloverRequest.setConditions(RolloverConditions.newBuilder(conditions).addMinIndexDocsCondition(1L).build());
         } else {
             rolloverRequest.setConditions(conditions);
         }
+        long currentMaxPrimaryShardDocs = rolloverRequest.getConditions().getMaxPrimaryShardDocs() != null
+            ? rolloverRequest.getConditions().getMaxPrimaryShardDocs()
+            : Long.MAX_VALUE;
+        if (targetIsTsdb && currentMaxPrimaryShardDocs > MAX_PRIMARY_SHARD_DOCS_FOR_TSDB) {
+            Map<String, Condition<?>> conditions = new HashMap<>(rolloverRequest.getConditions().getConditions());
+            conditions.put(MaxPrimaryShardDocsCondition.NAME, new MaxPrimaryShardDocsCondition(MAX_PRIMARY_SHARD_DOCS_FOR_TSDB));
+            rolloverRequest.setConditions(new RolloverConditions(conditions));
```

Comment:
Nit: we also have a builder that copies from an existing instance to simplify this code:
```
            rolloverRequest.setConditions(
                RolloverConditions.newBuilder(rolloverRequest.getConditions())
                    .addMaxPrimaryShardDocsCondition(MAX_PRIMARY_SHARD_DOCS_FOR_TSDB)
                    .build()
            );
```

Reply:
I think that `addMaxPrimaryShardDocsCondition(...)` fails if max primary doc count has already been defined? This can happen if it has been defined, but with a value higher than `MAX_PRIMARY_SHARD_DOCS_FOR_TSDB`.

Reply:
Hm.... ah damn I didn't think of it, I will add a note to fix it, that was not my intention with this builder :p. Thanks @martijnvg